pkg/tls/tls_test.go

```diff
@@ -0,0 +1,136 @@
+package tls
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"testing"
+	"time"
+)
+
+// generateTestCertificate creates a test certificate and private key for testing
+func generateTestCertificate(t *testing.T, notBefore, notAfter time.Time, dnsNames []string) ([]byte, []byte) {
+	// Generate a private key
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatalf("Failed to generate private key: %v", err)
+	}
+
+	// Create certificate template
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			Organization:  []string{"Test Org"},
+			Country:       []string{"US"},
+			Province:      []string{""},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{""},
+			PostalCode:    []string{""},
+		},
+		NotBefore:             notBefore,
+		NotAfter:              notAfter,
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		DNSNames:              dnsNames,
+	}
+
+	// Create the certificate
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	if err != nil {
+		t.Fatalf("Failed to create certificate: %v", err)
+	}
+
+	// Encode certificate to PEM
+	certPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certDER,
+	})
+
+	// Encode private key to PEM
+	privateKeyDER, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		t.Fatalf("Failed to marshal private key: %v", err)
+	}
+
+	keyPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privateKeyDER,
+	})
+
+	return certPEM, keyPEM
+}
+
+func TestParseTLSCertificate_Valid(t *testing.T) {
+	now := time.Now()
+	notBefore := now.Add(-24 * time.Hour)
+	notAfter := now.Add(24 * time.Hour)
+	dnsNames := []string{"example.com", "www.example.com"}
+
+	certPEM, keyPEM := generateTestCertificate(t, notBefore, notAfter, dnsNames)
```

Comment:
maybe a good idea to feed template into this function and then compare the parsed certificate from template which serves as the ground truth? 

Or better yet - we marshal a known/valid cert struct to bytes then use that initial cert struct as the ground truth and compare parsed result to that

Reply:
not hardcoded but I'm saying that the "actual" value (output from ParseTLSCertificate) that we compare should come from an explicitly defined expectation. 

Like if we define our expectation with x, why use f(x) for comparison? If f works as intended then x should contain all the information that we eventually want to use for comparison anyway